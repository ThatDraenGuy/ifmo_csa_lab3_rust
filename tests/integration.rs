use std::{
    error::Error,
    fs::{self, File},
    io::Read,
    path::Path,
};

use gag::BufferRedirect;
use insta::{assert_ron_snapshot, glob};
use lab3_rust::{machine, translator};
use serde::{Deserialize, Serialize};
use tempfile::{tempdir, TempDir};

#[derive(Deserialize)]
struct TestInput {
    translator_input: String,
    machine_input: String,
}

#[derive(Serialize)]
struct TestOutput {
    translator_output: String,
    stdout: String,
}

fn perform_test(temp_dir: &TempDir, input_path: &Path) -> Result<TestOutput, Box<dyn Error>> {
    // Достаём имя теста
    let test_name = input_path.file_stem().unwrap().to_str().unwrap();

    // Читаем входные данные
    let input_file = File::open(input_path)?;
    let input: TestInput = serde_yaml::from_reader(input_file)?;

    // Создаём временные файлы для входных и выходных данных
    let translator_input = temp_dir.path().join(test_name.to_owned() + "__translator_in");
    let machine_input = temp_dir.path().join(test_name.to_owned() + "__machine_in");
    let target = temp_dir.path().join(test_name.to_owned() + "__target");

    // Копируем входные данные
    fs::write(&translator_input, input.translator_input)?;
    fs::write(&machine_input, input.machine_input)?;

    // Хватаем вывод в STDOUT
    let mut stdout_buf = BufferRedirect::stdout()?;
    // Запускаем транслятор и виртуальную машину
    translator::main(&translator_input, &target)?;
    machine::main(&target, &machine_input)?;

    // Читаем выходные данные
    let mut stdout = String::new();
    stdout_buf.read_to_string(&mut stdout)?;

    // Читаем результат работы транслятора
    let translator_output = fs::read_to_string(target)?;

    //Возвращаем результат
    Ok(TestOutput { translator_output, stdout })
}

#[test]
fn test() {
    // Создаём одну временную директорию для всех файлов
    let dir = tempdir().unwrap();
    glob!("inputs/*.ron", |input_path| {
        // Проходимся по всем файлам с входными данными и для каждого из них создаём снапшот
        let result = perform_test(&dir, input_path).unwrap();
        assert_ron_snapshot!(result);
    })
}
