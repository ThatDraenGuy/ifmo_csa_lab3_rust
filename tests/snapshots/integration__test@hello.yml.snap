---
source: tests/integration.rs
expression: result
input_file: tests/inputs/hello.yml
---
translator_output: |-
  (
      code: [
          Data((5)),
          Data((72)),
          Data((101)),
          Data((108)),
          Data((108)),
          Data((111)),
          OpHigher(Math((
              opcode: Mov,
              args: RegImmed(Data, 0),
          ))),
          OpLower(MathOpLower(RegImmed(1))),
          OpHigher(Math((
              opcode: Mov,
              args: RegMemToReg(Count, Data),
          ))),
          OpHigher(Alter((
              opcode: Inc,
              arg: Data,
          ))),
          OpHigher(Math((
              opcode: Mov,
              args: RegMemToReg(Accumulator, Data),
          ))),
          OpHigher(Io((
              opcode: Out,
              arg: 1,
          ))),
          OpHigher(Alter((
              opcode: Dec,
              arg: Count,
          ))),
          OpHigher(Branch((
              opcode: Jnz,
              arg_higher: 0,
          ))),
          OpLower(BranchOpLower((
              arg_lower: 10,
          ))),
          OpHigher(Control((Exit))),
      ],
      entrypoint: (7),
  )
translator_result: OK
stdout: Hello
machine_result: OK
log: |
  [DEBUG lab3_rust::translator] Starting translation process
  [DEBUG lab3_rust::translator] Successfully finished translation process
  [DEBUG lab3_rust::machine] Starting virtual machine
  [DEBUG lab3_rust::machine::control_unit] TICK:    9; REGS: eax:   0, ecx:   0, edx:   0, eip:   8, esp:   0; INSTR: OpHigher(Math(MathOpHigher { opcode: Mov, args: RegImmed(Data, 0) }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   14; REGS: eax:   0, ecx:   0, edx:   1, eip:   9, esp:   0; INSTR: OpHigher(Math(MathOpHigher { opcode: Mov, args: RegMemToReg(Count, Data) }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   21; REGS: eax:   0, ecx:   5, edx:   1, eip:  10, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Inc, arg: Data }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   26; REGS: eax:   0, ecx:   5, edx:   2, eip:  11, esp:   0; INSTR: OpHigher(Math(MathOpHigher { opcode: Mov, args: RegMemToReg(Accumulator, Data) }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   33; REGS: eax:  72, ecx:   5, edx:   2, eip:  12, esp:   0; INSTR: OpHigher(Io(IoOpWord { opcode: Out, arg: 1 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   38; REGS: eax:  72, ecx:   5, edx:   2, eip:  13, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Dec, arg: Count }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   48; REGS: eax:  72, ecx:   4, edx:   2, eip:  15, esp:   0; INSTR: OpHigher(Branch(BranchOpHigher { opcode: Jnz, arg_higher: 0 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   52; REGS: eax:  72, ecx:   4, edx:   2, eip:  10, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Inc, arg: Data }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   57; REGS: eax:  72, ecx:   4, edx:   3, eip:  11, esp:   0; INSTR: OpHigher(Math(MathOpHigher { opcode: Mov, args: RegMemToReg(Accumulator, Data) }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   64; REGS: eax: 101, ecx:   4, edx:   3, eip:  12, esp:   0; INSTR: OpHigher(Io(IoOpWord { opcode: Out, arg: 1 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   69; REGS: eax: 101, ecx:   4, edx:   3, eip:  13, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Dec, arg: Count }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   79; REGS: eax: 101, ecx:   3, edx:   3, eip:  15, esp:   0; INSTR: OpHigher(Branch(BranchOpHigher { opcode: Jnz, arg_higher: 0 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   83; REGS: eax: 101, ecx:   3, edx:   3, eip:  10, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Inc, arg: Data }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   88; REGS: eax: 101, ecx:   3, edx:   4, eip:  11, esp:   0; INSTR: OpHigher(Math(MathOpHigher { opcode: Mov, args: RegMemToReg(Accumulator, Data) }))
  [DEBUG lab3_rust::machine::control_unit] TICK:   95; REGS: eax: 108, ecx:   3, edx:   4, eip:  12, esp:   0; INSTR: OpHigher(Io(IoOpWord { opcode: Out, arg: 1 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  100; REGS: eax: 108, ecx:   3, edx:   4, eip:  13, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Dec, arg: Count }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  110; REGS: eax: 108, ecx:   2, edx:   4, eip:  15, esp:   0; INSTR: OpHigher(Branch(BranchOpHigher { opcode: Jnz, arg_higher: 0 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  114; REGS: eax: 108, ecx:   2, edx:   4, eip:  10, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Inc, arg: Data }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  119; REGS: eax: 108, ecx:   2, edx:   5, eip:  11, esp:   0; INSTR: OpHigher(Math(MathOpHigher { opcode: Mov, args: RegMemToReg(Accumulator, Data) }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  126; REGS: eax: 108, ecx:   2, edx:   5, eip:  12, esp:   0; INSTR: OpHigher(Io(IoOpWord { opcode: Out, arg: 1 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  131; REGS: eax: 108, ecx:   2, edx:   5, eip:  13, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Dec, arg: Count }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  141; REGS: eax: 108, ecx:   1, edx:   5, eip:  15, esp:   0; INSTR: OpHigher(Branch(BranchOpHigher { opcode: Jnz, arg_higher: 0 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  145; REGS: eax: 108, ecx:   1, edx:   5, eip:  10, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Inc, arg: Data }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  150; REGS: eax: 108, ecx:   1, edx:   6, eip:  11, esp:   0; INSTR: OpHigher(Math(MathOpHigher { opcode: Mov, args: RegMemToReg(Accumulator, Data) }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  157; REGS: eax: 111, ecx:   1, edx:   6, eip:  12, esp:   0; INSTR: OpHigher(Io(IoOpWord { opcode: Out, arg: 1 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  162; REGS: eax: 111, ecx:   1, edx:   6, eip:  13, esp:   0; INSTR: OpHigher(Alter(AlterOpWord { opcode: Dec, arg: Count }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  172; REGS: eax: 111, ecx:   0, edx:   6, eip:  15, esp:   0; INSTR: OpHigher(Branch(BranchOpHigher { opcode: Jnz, arg_higher: 0 }))
  [DEBUG lab3_rust::machine::control_unit] TICK:  177; REGS: eax: 111, ecx:   0, edx:   6, eip:  16, esp:   0; INSTR: OpHigher(Control(ControlOpWord(Exit)))

